fatrace.c

```diff
@@ -275,7 +275,6 @@ print_event (const struct fanotify_event_metadata *data,
         close (proc_fd);
     } else {
         debug ("failed to open /proc/%i: %m", data->pid);
-        procname[0] = '\0';
```

Comment:
This isn't really a "cache", it's just a static string allocation because C. Like this it would just repeat the procname of the previous event where opening pid worked, which is just wrong. Please drop this commit.

This *could* be improved by moving to pidfds and keeping a pidfd → procname hash table cache. But that's much more work and definitively a separate PR.

Reply:
If the procname is wrong, then the PID will have changed, which is checked on line 282: `if (data->pid == procname_pid) {`. If it doesn't match, then the procname is deleted on line 287. Line 278 deleted the procname even if it was correct, and produced a lot of events with unknown procnames unnecessarily.

Reply:
Also, I read:

> A pidfd does not let you hold a reference to an individual thread, only to a process (or in kernel terminology, a thread group leader).
> A pidfd does not hold a reference to a pid number, nor does holding open a pidfd prevent the pid number of the underlying process from being reused.

https://www.corsix.org/content/what-is-a-pidfd

I'm not sure if that's promising.

Reply:
It is as `pid_t` is a crappy API -- getting information about the process via pid is inherently racy both because the process can go away (which you've seen already) and also because pids get recycled. But indeed there is no better API to get from a pidfd to `cmdline` and/or `exe`, so duh.. I suppose the only race free way is EBPF then. I [dabbled with that a bit a few years ago](https://piware.de/post/2020-02-28-bpftrace/), maybe that's worth another look..

Reply:
That looks a lot more promising. The racy API is a fundamentally broken foundation for what fatrace tries to accomplish, so there will be a lot of complex code that only partly duck-tapes over the holes. It seems BPF and C can speak, so formatting shouldn't be a problem.